File: Stack.py
class Node:
    def __init__(self,data = None):
        self.data = data
        self.next = None


class stack:
    def __init__(self): 
        self.top = None
        self.size = 0

    def push(self,data): #Push operation
        node = Node(data)
        if self.top:
            node.next = self.top
            self.top = node
        else:
            self.top = node
        self.size += 1

    def pop(self):  #Pop operation
        if self.top:
            data = self.top.data
            self.size -= 1
            if self.top.next:
                self.top = self.top.next
            else:
                self.top = None
                return data
        else:
            return None

File: Stack.rs
struct Node {
    value: i32,
    next: Option<Box<Node>>,
}

struct Stack {
    top: Option<Box<Node>>,
    size: i8,
}

impl Stack {
    fn new() -> Stack {
        Stack { top: None, size: 0 }
    }

    fn push(&mut self, value: i32) {
        let new_node = Box::new(Node {
            value,
            next: self.top.take(),
        });
        self.top = Some(new_node);
        self.size += 1;
    }

    fn print_stack(self) {
        let mut current = self.top;
        while let Some(node) = current {
            println!("{}", node.value);
            current = node.next;
        }
    }

    fn pop(&mut self) -> Option<i32> {
        match self.top.take() {
            Some(boxed_node) => {
                let node = *boxed_node; // Extract the Node from the Box
                self.top = node.next; // Update top to point to the next node
                self.size -= 1; // Decrease the size of the stack
                Some(node.value) // Return the value of the popped Node
            }
            None => None, // Stack is empty
        }
    }
}

fn main() {
    let mut stack = Stack::new();
    stack.push(1);
    stack.push(2);
    stack.push(3);
    stack.pop();
    stack.print_stack();
}
